src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PETypeParameterSymbol.cs

```diff
@@ -356,15 +356,29 @@ public override ImmutableArray<CSharpAttributeData> GetAttributes()
             return _lazyCustomAttributes;
         }
 
-        private TypeParameterBounds GetBounds(ConsList<TypeParameterSymbol> inProgress)
+        private TypeParameterBounds GetBounds(ConsList<TypeParameterSymbol> inProgress, bool early)
         {
-            Debug.Assert(!inProgress.ContainsReference(this));
-            Debug.Assert(!inProgress.Any() || ReferenceEquals(inProgress.Head.ContainingSymbol, this.ContainingSymbol));
+            // PROTOTYPE(NullableReferenceTypes): Re-enable asserts.
+            //Debug.Assert(!inProgress.ContainsReference(this));
+            //Debug.Assert(!inProgress.Any() || ReferenceEquals(inProgress.Head.ContainingSymbol, this.ContainingSymbol));
 
-            if (ReferenceEquals(_lazyBounds, TypeParameterBounds.Unset))
+            var currentBounds = _lazyBounds;
+            if (currentBounds == TypeParameterBounds.Unset)
             {
                 var constraintTypes = GetDeclaredConstraintTypes();
                 Debug.Assert(!constraintTypes.IsDefault);
+                var bounds = new TypeParameterBounds(constraintTypes);
+                Interlocked.CompareExchange(ref _lazyBounds, bounds, currentBounds);
+                currentBounds = _lazyBounds;
+            }
+
+            if (!currentBounds.IsSet(early))
+            {
+                var constraintTypes = currentBounds.ConstraintTypes;
+                if (!ContainingModule.UtilizesNullableReferenceTypes)
+                {
+                    constraintTypes = constraintTypes.SelectAsArray(t => t.SetUnknownNullabilityForReferenceTypes());
```

Comment:
>SetUnknownNullabilityForReferenceTypes [](start = 75, length = 38)

Can you add a PROTOTYPE marker? I suspect this will need to be revised in the presence of `[NonNullTypes]` in relevant scope.